pkg/etcdcli/health.go

```diff
@@ -41,66 +42,28 @@ type healthCheck struct {
 type memberHealth []healthCheck
 
 func getMemberHealth(ctx context.Context, etcdMembers []*etcdserverpb.Member) memberHealth {
-	var wg sync.WaitGroup
 	memberHealth := memberHealth{}
-	hch := make(chan healthCheck, len(etcdMembers))
 	for _, member := range etcdMembers {
 		if !HasStarted(member) {
 			memberHealth = append(memberHealth, healthCheck{Member: member, Healthy: false})
 			continue
 		}
-		wg.Add(1)
-		go func(member *etcdserverpb.Member) {
-			defer wg.Done()
-			// If the endpoint is for a learner member then we should skip testing the connection
-			// via the member list call as learners don't support that.
-			// The learner's connection would get tested in the health check below
-			skipConnectionTest := false
-			if member.IsLearner {
-				skipConnectionTest = true
-			}
-			cli, err := newEtcdClientWithClientOpts([]string{member.ClientURLs[0]}, skipConnectionTest)
-			if err != nil {
-				hch <- healthCheck{Member: member, Healthy: false, Error: fmt.Errorf("create client failure: %w", err)}
-				return
-			}
-			defer func() {
-				if err := cli.Close(); err != nil {
-					klog.Errorf("error closing etcd client for getMemberHealth: %v", err)
-				}
-			}()
-
-			st := time.Now()
-			ctx, cancel := context.WithCancel(ctx)
-			var resp *clientv3.GetResponse
-			if member.IsLearner {
-				// Learner members only support serializable (without consensus) read requests
-				resp, err = cli.Get(ctx, "health", clientv3.WithSerializable())
-			} else {
-				// Linearized request to verify health of a voting member
-				resp, err = cli.Get(ctx, "health")
-			}
-
-			hc := healthCheck{Member: member, Healthy: false, Took: time.Since(st).String()}
-			if err == nil {
-				if resp.Header != nil {
-					raftTerms.Set(member.Name, resp.Header.RaftTerm)
-				}
-				hc.Healthy = true
-			} else {
-				klog.V(2).Infof("health check for memberID (%v) failed: err(%v)", resp.Header.MemberId, err)
-				hc.Error = fmt.Errorf("health check failed: %w", err)
-			}
-			cancel()
-			hch <- hc
-		}(member)
-	}
 
-	wg.Wait()
-	close(hch)
+		resChan := make(chan healthCheck, 1)
```

Comment:
resChan is never closed, might not matter since the GC will eat it after scope end

Reply:
good catch, TIL that it's actually also not necessary in this context:

> In Go, the channel is not like the resources like file descriptor. It does not have to be closed, it is just an object, close (channel) is only used to tell the receivers that there is nothing to write, no other purpose.

https://faun.pub/implmenting-timeout-in-golang-ee2bc4aa6ae4

Reply:
added the close and put the cancel on top, as I just saw the picker_wrapper.go implementation will cancel based on that:
https://github.com/grpc/grpc-go/blob/v1.40.0/picker_wrapper.go#L103-L117

which is where we were actually stuck somehow... I actually thought the deadline exceeded must've kicked in earlier here.. This was literally hanging for half an hour...